### PR TITLE
feat(schema, api-client, api, cli): Add api-key schemas and types; Fix schema inconsistencies; Minor fix for CLI build errors 

### PR DESCRIPTION
**Pull request by muntaxir4**
## Description

**`@keyshade/schema`**
- Add schemas and types for `api-key`.
- Add tests for `api-key` schemas.
- Export api-key schemas from `index` and its types  from `index.types` .
- Fixed inconsistencies between schemas and actual types in API.
- Changed date properties to `z.string().datetime()` for stricter validation.

**`@keyshade/api-client`**
- Implement the `api-key` Controller and its tests.

**`api`**
- Implement pagination response for `getApiKeysOfUser`.

**`cli`**
- Fix some build errors caused due to schemas.
<br />
Related to #519 

## Future Improvements

CLI build errors could be fixed for secrets and variables.

## Developer's checklist

- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-check on my work

**If changes are made in the code:**

- [x] I have followed the [coding guidelines](https://google.github.io/styleguide/jsguide.html)
- [x] My changes in code generate no new warnings
- [ ] My changes are breaking another fix/feature of the project
- [x] I have added test cases to show that my feature works
- [ ] I have added relevant screenshots in my PR
- [x] There are no UI/UX issues

### Documentation Update

- [ ] This PR requires an update to the documentation at [docs.keyshade.xyz](https://docs.keyshade.xyz)
- [x] I have made the necessary updates to the documentation, or no documentation changes are required.

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Added comprehensive schemas for API key operations using zod, including request and response schemas.
- Defined and exported TypeScript types inferred from the API key schemas.
- Updated the main index files to export the new API key schemas and types.
- Implemented extensive tests for the API key schemas, covering both valid and invalid scenarios.
- Adjusted the tsconfig.json to correct the base path extension.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>index.ts</strong><dd><code>Add comprehensive API key schemas and transformations</code>&nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/schema/src/api-key/index.ts

<li>Added comprehensive API key schemas using zod.<br> <li> Defined request and response schemas for API key operations.<br> <li> Included transformations for date fields.<br>

</details>

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/557/files#diff-eb680e1da49b1fb5b8806fe44f2a7cafcbef3240be6a1ea689f865b58bb363ec">+64/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>index.types.ts</strong><dd><code>Define and export API key types</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/schema/src/api-key/index.types.ts

<li>Defined TypeScript types inferred from API key schemas.<br> <li> Exported types for API key operations.<br>

</details>

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/557/files#diff-4f87da4b01ced7bfa38e59b0c3be15dbd3e6e0970da86c83f99aba4d2e76697a">+50/-0</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>index.ts</strong><dd><code>Export API key schemas from index</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/schema/src/index.ts

- Exported API key schemas from the main index file.

</details>

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/557/files#diff-5c37949968eac79dd5d1c0df1176a9073fd624d2be2742d3c7d718e318aa2cdc">+2/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>index.types.ts</strong><dd><code>Export API key types from index.types</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/schema/src/index.types.ts

- Exported API key types from the main index types file.

</details>

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/557/files#diff-47f389a66e4515c1526d27bc56ae56141418a0a0419271b98077614f7778d3e5">+1/-4</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>api-key.spec.ts</strong><dd><code>Add tests for API key schemas</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/schema/tests/api-key.spec.ts

<li>Added tests for API key schemas.<br> <li> Validated both valid and invalid cases for each schema.<br>

</details>

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/557/files#diff-1fc360b5dc6d059c5ea0d9e3eeb3ff6f0f7e7b888777e12e60d90e6b487c07ae">+292/-8</a>&nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Configuration changes</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>tsconfig.json</strong><dd><code>Update tsconfig base path</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

packages/schema/tsconfig.json

- Modified the path for extending base tsconfig.

</details>

  </td>
  <td><a href="https://github.com/keyshade-xyz/keyshade/pull/557/files#diff-fc763b82b8113516d39db871392775c27b0e8427331bd7f0add8c64d3101802b">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

**Comment by muntaxir4** [comment]
I will try to modify the date types for previously implemented schemas and also update the docs for integration in this PR.

**Review by rajdip-b — COMMENTED** [review]
Looks good, can you also include the API client changes?

**Comment by muntaxir4** [comment]
> Looks good, can you also include the API client changes?

Implementation for API-KEY controller? I was thinking of creating a new PR for that.

**Comment by rajdip-b** [comment]
Actually the parent issue is a centralized one, so i was thinking it would be best to stash both of the changes in here

**Comment by muntaxir4** [comment]
> Actually the parent issue is a centralized one, so i was thinking it would be best to stash both of the changes in here

Done with this.

File: packages/schema/src/environment/index.ts [hunk 1]
```diff
@@ -6,14 +6,14 @@ export const EnvironmentSchema = z.object({
...
-  updatedAt: z.string(),
+  updatedAt: z.string().datetime(),
+  createdAt: z.string().datetime(),
   lastUpdatedById: z.string(),
   projectId: z.string()
 })
 
 export const CreateEnvironmentRequestSchema = z.object({
-  name: z.string(),
+  name: EnvironmentSchema.shape.name,
```
Comment (rajdip-b) [inline packages/schema/src/environment/index.ts:16]: Any special reason why you are using this construct? It's correct but i feel its unnecessary. And if we are to use this construct, then we should use it everywhere.

File: packages/schema/src/api-key/index.ts [hunk 2]
```diff
@@ -0,0 +1,61 @@
+import { z } from 'zod'
+import { expiresAfterEnum, authorityEnum } from '@/enums'
+import { PageRequestSchema } from '@/pagination'
+
+export const ApiKeySchema = z.object({
+  id: z.string(),
+  name: z.string(),
+  slug: z.string(),
+  value: z.string(),
+  expiresAt: z.string().datetime().nullable(),
```
Comment (rajdip-b) [inline packages/schema/src/api-key/index.ts:10]: This should be `expiresAfterEnum`

File: packages/schema/src/api-key/index.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  createdAt: z.string().datetime(),
+  updatedAt: z.string().datetime(),
+  authorities: z.array(authorityEnum),
+  userId: z.string()
+})
+
+export const CreateApiKeyRequestSchema = z.object({
+  name: ApiKeySchema.shape.name,
+  expiresAfter: expiresAfterEnum.optional(),
```
Comment (rajdip-b) [inline packages/schema/src/api-key/index.ts:19]: Derive this from the parent shape and make it optional

**Review by rajdip-b — CHANGES_REQUESTED** [review]

File: packages/schema/src/api-key/index.ts (hunk 2 above)
Comment (muntaxir4) [reply packages/schema/src/api-key/index.ts:10]: The prisma schema has DateTime as type

Comment (rajdip-b) [reply packages/schema/src/api-key/index.ts:10]: oh righ yes

File: packages/schema/src/api-key/index.ts (hunk 3 above)
Comment (muntaxir4) [reply packages/schema/src/api-key/index.ts:19]: Doing this, however the dto for create end point had only these properties.

**Review by rajdip-b — APPROVED** [review]

**Review by rajdip-b — APPROVED** [review]

**Pull request merged**